src/walk.rs

```diff
@@ -205,12 +207,8 @@ fn spawn_receiver(
                 }
 
                 // Wait for all threads to exit before exiting the program.
-                let mut results: Vec<ExitCode> = Vec::new();
-                for h in handles {
-                    results.push(h.join().unwrap());
-                }
-
-                merge_exitcodes(&results)
+                let results = handles.into_iter().map(|h| h.join().unwrap());
+                merge_exitcodes(results)
```

Comment:
:+1:

Reply:
Actually, I now noticed that this is semantically different: Since `Iterator::any` stops early once any of the values satisfies the predicate, the later threads won't be joined if one of them returned an error. I assume this should not happen.

To avoid the `collect`, I could add one of two things: A `exhaustive_any` extension to iterators that fully consumes the iterator (so every join call happens), or a `ScopeThread` that is joined once the scope of the handle ends / the handle is dropped. I think the first option would be simpler and easier to read.

Ofc, we could also collect into a vector again, but that would not be very satisfying :laughing:

Reply:
Update: Actually, we can also just guard the iterator to make sure it is fully exhausted. Here is an example:

```rust
struct LoudNumber(i32);
impl Drop for LoudNumber {
    fn drop(&mut self) {
        println!("Dropping {}", self.0);
    }
}
fn main() {
    let mut iter = (0..10).map(LoudNumber);
    let _ = iter.any(|x| x.0 == 5);

    println!("---");
    let mut iter = scopeguard::guard((0..10).map(LoudNumber), |remaining_items| {
        remaining_items.for_each(drop)
    });
    let _ = iter.any(|x| x.0 == 5);
}
```

This outputs:
```console
Dropping 0
Dropping 1
Dropping 2
Dropping 3
Dropping 4
Dropping 5
---
Dropping 0
Dropping 1
Dropping 2
Dropping 3
Dropping 4
Dropping 5
Dropping 6
Dropping 7
Dropping 8
Dropping 9
```

What do you think?

Reply:
Personally I'd prefer the explicit collect over the extra dependency.  `for_each(drop)` is misleading IMO, since it's not like the items wouldn't otherwise be dropped.  This is more straightforward IMO:

```rust
                // Wait for all threads to exit before exiting the program.
                let results: Vec<_> = handles
                    .into_iter()
                    .map(|h| h.join().unwrap())
                    .collect();

                merge_exitcodes(results)
```

Reply:
I agree that `for_each(drop)` to force side effects in `map` is more involved than this code should be and new readers probably have to think about the code rather than just seeing that we just want to join all threads. I saw some discussion about an `Iterator::exhaust(self)` or similar in Rust forums and issues that would be equivalent to `for_each(drop)`, but it got turned down because the latter already exists.

We could do two things here:
1. Don't change the code and collect into vector.
2. Add an extension trait with `.exhausted` or similar that returns an adapter that consumes the remaining iterator on drop. This could increase readability in callers code (and you could hover the method to see what it is supposed to do). I think I will check if this option is feasible and maybe push a suggestion here. (we could also have this in another crate if you like)

Honestly, I would be fine with both approaches though. If we go with the first approach, the change to the `merge_exitcodes` signature doesn't make sense until msrv is up to 1.53 with `[T; N]: IntoIterator`. I would just revert and add a `FIXME:` comment referring to this PR.